src/DurableTask.SqlServer/Scripts/logic.sql

```diff
@@ -1252,3 +1266,138 @@ BEGIN
+                                                           FROM History f
+                                                           WHERE f.[TaskHub] = @TaskHub AND f.[InstanceID] = @InstanceID  AND f.[TaskID] = h.[TaskID] AND f.[EventType] = CASE WHEN h.[EventType] = 'TaskScheduled' THEN 'TaskFailed' ELSE 'SubOrchestrationInstanceFailed' END)))
+
+    -- Mark all events related to failure as rewound
+    UPDATE Payloads
+    SET [Reason] = CONCAT('Rewound: ', ef.[EventType])
+    FROM Payloads p
+    JOIN @eventsInFailure ef ON p.[PayloadID] = ef.[DataPayloadID]
+    WHERE [TaskHub] = @TaskHub AND [InstanceID] = @InstanceID AND [SequenceNumber] IN (SELECT [SequenceNumber] FROM @eventsInFailure WHERE [DataPayloadID] IS NOT NULL AND [EventType] IN ('TaskScheduled', 'SubOrchestrationInstanceCreated'))
+
+    DECLARE @sequenceNumber bigint
+    DECLARE @eventType varchar(40)
+    DECLARE @payloadId uniqueidentifier
+    DECLARE sequenceNumberCursor CURSOR LOCAL FOR
+        SELECT [SequenceNumber], [EventType]
+        FROM @eventsInFailure
+        WHERE [DataPayloadID] IS NULL
+
+    OPEN sequenceNumberCursor 
+    FETCH NEXT FROM sequenceNumberCursor INTO @sequenceNumber, @eventType
+
+    WHILE @@FETCH_STATUS = 0 BEGIN
+        SET @payloadId = NEWID()
+        INSERT INTO Payloads (
+            [TaskHub],
+            [InstanceID],
+            [PayloadID],
+            [Reason]
+        )
+        VALUES (@TaskHub, @InstanceID, @payloadId, CONCAT('Rewound: ', @eventType))
+        FETCH NEXT FROM sequenceNumberCursor INTO @sequenceNumber, @eventType
+    END
+    CLOSE sequenceNumberCursor
+    DEALLOCATE sequenceNumberCursor
+
+    -- Transform all events related to failure to GenericEvents, except for SubOrchestrationInstanceStarted that can be kept
+    UPDATE History
+    SET [EventType] = 'GenericEvent'
+    WHERE [TaskHub] = @TaskHub AND [InstanceID] = @InstanceID
+      AND ([SequenceNumber] IN (SELECT [SequenceNumber]
+                                FROM @eventsInFailure WHERE [EventType] <> 'SubOrchestrationInstanceStarted')
+           OR [RuntimeStatus] = 'Failed')
+
+    -- Enumerate instances of sub orchestrations related to SubOrchestrationInstanceFailed events and rewing them recursively
+    DECLARE @subOrchestrationInstanceID varchar(100)
+    DECLARE subOrchestrationCursor CURSOR LOCAL FOR
+        SELECT i.[InstanceID]
+        FROM dt.Instances i
+          JOIN dt.History h ON i.[TaskHub] = h.[TaskHub] AND i.[InstanceID] = h.[InstanceID]
+          JOIN @eventsInFailure e ON e.[TaskID] = h.[TaskID]
+        WHERE i.[ParentInstanceID] = @InstanceID 
+          AND h.[EventType] = 'ExecutionStarted'
+          AND e.[EventType] = 'SubOrchestrationInstanceFailed'
+
+
+    OPEN subOrchestrationCursor 
+    FETCH NEXT FROM subOrchestrationCursor INTO @subOrchestrationInstanceID
+
+    WHILE @@FETCH_STATUS = 0 BEGIN
+        EXECUTE dt._RewindInstance @subOrchestrationInstanceID, @Reason
```

Comment:
Should we add a comment here about the recursive depth limit?

Reply:
I'm actually wondering if we could get rid of the recursion by collecting all instances with failures in the dependency tree at once. Do you think it is worth checking the feasibility ? Or maybe in a second change afterwards ?